### PR TITLE
Integration Test Framework

### PR DESCRIPTION
**Pull request by muralidharm**
This is a pull request to add the framework for integration tests for the CDF Fabric replicator.  These tests do not currently pass, but generate timeseries data to send to cdf and check a specified lakehouse destination to make the comparison.  The next step is to configure a test replicator.
To run these tests, copy `.env.example` and create an `.env` file, filling in the value.  Uncomment line 19 in `test_integration.py` to disable the skip on the test.
If a lakehouse table with the proper schema does not exist, create one using the following code:
```
    credential = DefaultAzureCredential()
    df = pd.DataFrame(
        {
            "externalId": ["hist:mtu:39tic1062.pv"],
            "timestamp": ["2022-02-07T16:03:47.000Z"],
            "value": [19.338623046875],
        }
    )
    token = credential.get_token("https://storage.azure.com/.default")

    path = "<Lakehouse Path>/Tables/timeseries"

    write_deltalake(path, df, mode="append", storage_options={"bearer_token": token.token, "use_fabric_endpoint": "true"})
    print("Wrote to Delta Lake" + df.to_string())
```

### CODE REVIEW

File: tests/integration/conftest.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
+import pytest
+import os
+from dotenv import load_dotenv
+from cognite.client import ClientConfig, CogniteClient
+from cognite.client.credentials import OAuthClientCredentials
+from azure.identity import DefaultAzureCredential
+from tests.integration.integration_steps.cdf_steps import remove_time_series_data
+from tests.integration.integration_steps.time_series_generation import generate_datapoints, generate_timeseries_set
```
Comment (kenny-chalupa) [inline tests/integration/conftest.py]: `generate_datapoints` isn't used

File: tests/integration/integration_steps/cdf_steps.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
+import pandas as pd
+from datetime import datetime
+from time import sleep
+from cognite.client import CogniteClient
+from cognite.client.data_classes import Datapoint, DatapointsArray, TimeSeries, TimeSeriesWrite, TimeSeriesList
```
Comment (kenny-chalupa) [inline tests/integration/integration_steps/cdf_steps.py]: `DataPointsArray` and `TimeSeriesList` are not used as well

**Review by kenny-chalupa — APPROVED** [review]
small comments for unessesary imports but otherwise looks good

File: tests/integration/conftest.py (hunk 1 above)
Comment (muralidharm) [reply tests/integration/conftest.py]: Removed

File: tests/integration/integration_steps/cdf_steps.py (hunk 2 above)
Comment (muralidharm) [reply tests/integration/integration_steps/cdf_steps.py]: Removed

File: .env.example [hunk 3]
```diff
@@ -7,3 +7,7 @@ COGNITE_PROJECT=
 COGNITE_BASE_URL=
 COGNITE_STATE_DB=
 COGNITE_STATE_TABLE=
+COGNITE_EXTRACTION_PIPELINE=
+LAKEHOUSE_BASE_URL=
+LAKEHOUSE_TABLE_NAME=
+LAKEHOUSE_ABFSS_PREFIX=
```
Comment (wbevans) [inline .env.example:13]: As discussed in Teams, should probably converge on using either BASE_URL or ABFSS_PREFIX, but having both is confusing.

**Review by wbevans — APPROVED** [review]

File: tests/integration/conftest.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+@pytest.fixture(scope="session")
+def azure_credential():
+    return DefaultAzureCredential()
+
+@pytest.fixture(scope="session")
+def lakehouse_timeseries_path(azure_credential):
+    lakehouse_timeseries_path = os.environ["LAKEHOUSE_ABFSS_PREFIX"] + "/Tables/" + os.environ["DPS_TABLE_NAME"]
+    yield lakehouse_timeseries_path
+    delta_table = get_ts_delta_table(azure_credential, lakehouse_timeseries_path)
```
Comment (xtophs) [inline tests/integration/conftest.py:55]: Are those 2 lines in scope for a function lakehouse_timeseries_path() ... at first glance it looks like the function does more than return the path

File: tests/integration/integration_steps/cdf_steps.py [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+    cognite_client.time_series.create(time_series_write_list)
+    return time_series_data
+
+def push_data_to_cdf(time_series_data: list[TimeSeries], cognite_client: CogniteClient):
+    time_series_data_points_pushed = {}
+    for ts in time_series_data:
+        time_series_data_points_pushed[ts.external_id] = push_data_points_to_cdf(
+            data_points=ts.datapoints, external_id=ts.external_id, cognite_client=cognite_client
+        )
+    sleep(5)
```
Comment (xtophs) [inline tests/integration/integration_steps/cdf_steps.py:56]: Why is that needed?

File: tests/integration/integration_steps/cdf_steps.py [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+    # Create a data model in CDF
+    pass
+
+def update_data_model_in_cdf():
+    # Update a data model in CDF
+    pass
+
+
+def compare_timestamps(timestamp1: datetime, timestamp2: datetime) -> bool:
+    return timestamp1.replace(microsecond=0) == timestamp2.replace(microsecond=0)
```
Comment (xtophs) [inline tests/integration/integration_steps/cdf_steps.py:74]: why are we ignoring microseconds in the comparison? Maybe add a comment why this is needed?

File: tests/integration/integration_steps/fabric_steps.py [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+def get_ts_delta_table(credential: DefaultAzureCredential, lakehouse_timeseries_path: str):
+    token = credential.get_token("https://storage.azure.com/.default")
+    return DeltaTable(lakehouse_timeseries_path,storage_options={"bearer_token": token.token, "use_fabric_endpoint": "true"},)
+
+def read_deltalake_timeseries(timeseries_path:str, credential: DefaultAzureCredential):
+    delta_table = get_ts_delta_table(credential, timeseries_path)
+    df = delta_table.to_pandas()
+    return df
+
+def prepare_lakehouse_dataframe_for_comparison(dataframe, external_id):
```
Comment (xtophs) [inline tests/integration/integration_steps/fabric_steps.py:19]: some functions define return types, others don't. What's the criteria to determine if a return type is explicitly defined?

File: tests/integration/integration_steps/cdf_steps.py (hunk 6 above)
Comment (muralidharm) [reply tests/integration/integration_steps/cdf_steps.py:74]: The microsecond precision is lost to the thousandths of milliseconds at some point in the process so we're rounding to seconds.  I'll add a comment

File: tests/integration/integration_steps/cdf_steps.py (hunk 5 above)
Comment (muralidharm) [reply tests/integration/integration_steps/cdf_steps.py:56]: This allows time for the data points to be pushed to CDF before moving on to the next step in the test.  I preferred doing this here than in the test itself

File: tests/integration/conftest.py (hunk 4 above)
Comment (muralidharm) [reply tests/integration/conftest.py:55]: I don't want to return the delta table here since I need to make the read after the data has been pushed, but I also want to clean up the table after the test.

File: tests/integration/integration_steps/fabric_steps.py (hunk 7 above)
Comment (muralidharm) [reply tests/integration/integration_steps/fabric_steps.py:19]: We should probably align with Cognite, who seem to explicitly define return types.  Was an oversight on my part, I haven't done this on previous projects I worked on in Python

File: tests/integration/integration_steps/cdf_steps.py (hunk 5 above)
Comment (xtophs) [reply tests/integration/integration_steps/cdf_steps.py:56]: is there a more deterministic way to ensure that the push completed before moving on? This might turn into a random, hard to detect test failure whenever the 5 seconds are not enough

Comment (kenny-chalupa) [reply tests/integration/integration_steps/cdf_steps.py:56]: I had introduced listening to the subscription in our first round of integration testing and that was automatic without  having to wait that may be enough if we want to use that
